Network stall detection

avelad opened the pull request:
Closes: https://github.com/google/shaka-player/issues/1582

TheModMaker commented:
Joey's suggestion in that issue was if we went a certain time without getting more data, we'd stall.  This cancels the timer after the first progress, so this is basically a connection timeout.  If you want that, I'd suggest naming it `connectionTimeout` instead, or change to reset the timer on each progress to keep it a stall timeout.

Also, since the code is basically the same between the two handlers, I'd suggest moving to NetworkingEngine.  This also allows app-provided handlers to get this too without having to implement it.

avelad commented:
I am going to implement both suggestions. Thanks!

avelad commented:
I have tried to move it to NetworkingEngine, but if I abort the operation from there, it had an unwanted recursion, I have not managed to fix it, so in the end I have left it in each plugin.

I have fixed the use of stallTimeout and added connectionTimeout.

@TheModMaker can you review again?

caridley commented on lib/net/http_fetch_plugin.js:
```diff
@@ -54,8 +63,38 @@ shaka.net.HttpFetchPlugin = class {
+
+    /** @type {shaka.extern.ProgressUpdated} */
+    const onProgressUpdated = (time, bytes, numBytesRemaining) => {
+      if (stallTimer) {
+        stallTimer.stop();
+      }
```
Should this be the connectionTimer ?

avelad replied:
Fixed!

caridley commented on lib/net/http_fetch_plugin.js:
```diff
@@ -69,20 +108,25 @@ shaka.net.HttpFetchPlugin = class {
+      if (stallTimer) {
+        stallTimer.stop();
+      }
+    });
```
should the connectionTimer be stopped as well?

avelad replied:
Fixed!

avelad commented:
This is really very beneficial, after testing internally for hours, we have seen that this actually makes buffering less times, so the user experience is better. Note: in our case 'streaming.retryParameters.maxAttempts' is 15.

avelad commented:
@TheModMaker can you review again?

joeyparrish commented on docs/tutorials/network-and-buffering-config.md:
````diff
@@ -15,6 +15,8 @@ identical:
 ```js
 retryParameters: {
   timeout: 0,       // timeout in ms, after which we abort; 0 means never
+  stallTimeout: 0,  // stall timeout in ms, after which we abort; 0 means never
````
Can you set these in the docs to their default values, please?

avelad replied:
If I understand you correctly, you want the default value in the ShakaPlayer configuration to use 0 for both new configurations, is that correct?

joeyparrish replied:
No, sorry.  I mean that the docs should show the default values.  You gave non-zero defaults, and I agree with those defaults.

avelad replied:
Sorry for taking time to respond, Friday was bank holiday in Spain. I fix it now.

joeyparrish commented on lib/net/http_fetch_plugin.js:
```diff
@@ -54,8 +63,38 @@ shaka.net.HttpFetchPlugin = class {
+        connectionTimer.stop();
+      }
+      if (stallTimer) {
+        stallTimer.stop();
```
tickAfter implies a reset, so you can drop the stop() call here.

avelad replied:
Done!

joeyparrish commented on lib/net/http_xhr_plugin.js:
```diff
@@ -86,7 +98,28 @@ shaka.net.HttpXHRPlugin = class {
+          connectionTimer.stop();
+        }
+        if (stallTimer) {
+          stallTimer.stop();
```
Same here.  You can drop the stop() call if you're calling tickAfter().

avelad replied:
Done!

joeyparrish commented on test/net/networking_engine_unit.js:
```diff
@@ -90,6 +90,8 @@ describe('NetworkingEngine', /** @suppress {accessControls} */ () => {
         backoffFactor: 0,
         fuzzFactor: 0,
         timeout: 0,
+        stallTimeout: 0,
```
Would a new test case make sense?

avelad replied:
I valued it, but the problem is that there is no timeout test to be able to compare :S

joeyparrish replied:
I understand.  It's tough sometimes to figure out new test cases without a similar example to work from.  Please file an issue to follow-up on this with tests, and we'll try to add some later.  Don't worry about it for now.  The code looks clear and short enough to me.  (I hope I don't regret saying that some day!  😛)

avelad replied:
Once merged I create the issue and link it here.

joeyparrish reviewed (changes requested):
I would also like @TheModMaker's opinion again on doing this generically in NetworkingEngine vs in each plugin.  If the plugin produces progress events, I feel as if you could do it generically.  If the plugin does not produce progress events, though, it may always trigger connection or stall timeouts.  Could NetworkingEngine tell which plugins would generate those events?  My instinct is "no", but I haven't analyzed it.

avelad commented:
@joeyparrish I have moved the logic to `NetworkingEngine`. There is nothing that is exported to know if the plugin emits progress events or not, but since both `xhr` and `fetch` emit it, I think that right now it is not necessary to add this logic.

joeyparrish commented:
> There is nothing that is exported to know if the plugin emits progress events or not, but since both `xhr` and `fetch` emit it, I think that right now it is not necessary to add this logic.

That could still cause problems for custom plugins.  For example, there are apps and other libraries which plug into the networking engine in order to load content over peer-to-peer connections.  If those have not implemented progress events, we would see them as having stalled, correct?

Here's a nice compromise:  We can keep the centralized code for stalls and timeouts, and add a flag to network plugins to declare that they do emit progress events.  Then you can add this flag to the fetch & XHR plugins.  This would allow us to have this feature without breaking existing plugins.

What do you think?

joeyparrish reviewed (commented):
Looks good apart from the need for a flag to avoid breaking old plugins

avelad commented:
@joeyparrish I have already added what you have commented, can you review again?

joeyparrish commented on lib/net/networking_engine.js:
```diff
@@ -429,6 +431,7 @@ shaka.net.NetworkingEngine = class extends shaka.util.FakeEventTarget {
               shaka.util.Error.Code.UNSUPPORTED_SCHEME,
               uri));
     }
+    const progressSupport = object ? object.progressSupport : false;
```
I think this can be object.progressSupport.  By this point in the code, you know object is non-null, right?  Or does the compiler not understand that detail?

avelad replied:
Sometimes I want to be too purist, sorry, now I change it.

joeyparrish commented on lib/net/http_fetch_plugin.js:
```diff
@@ -229,6 +229,15 @@ shaka.net.HttpFetchPlugin = class {
   }
+
+  /**
+   * Determine if progress event is supported.
```
In these particular plugins, there is no runtime determination to be made.  So I think these methods are not needed.  You can just pass `true` to the corresponding argument of `registerScheme()`.

avelad replied:
ok!

joeyparrish reviewed (changes requested):
Thanks for taking care of this!

avelad commented:
@joeyparrish Do you think you could have time today to approve and merge this PR?

avelad commented:
@joeyparrish can you run the Shaka-bot?

michellezhuogg commented:
> @joeyparrish can you run the Shaka-bot?

Running that now :)

avelad commented:
Thanks @michellezhuogg !
